File: client/TwendeZanzabar/TwendeZanzabar/Auth/AuthController.swift
import AuthenticationServices
import Combine
import CryptoKit
import Foundation
import SwiftUI

enum AuthStatus { case unknown, signedOut, signingIn, signedIn, error }

@MainActor
final class AuthController: NSObject, ObservableObject {
    @Published var status: AuthStatus = .unknown
    @Published var errorMessage: String?

    private let storage = AuthStorage.shared

    var session: AuthStorage.Session? { storage.session }

    func bootstrap() {
        status = storage.session == nil ? .signedOut : .signedIn
    }

    /// Dev-only: create a local preview session without any auth server.
    func signInAsDev() {
        let user = UserInfo(
            id: "00000000-0000-0000-0000-000000000000",
            email: "preview@twende.local",
            full_name: "Preview Traveler",
            role: "traveler",
            phone: nil
        )
        storage.write(
            user: user,
            access: "dev-bypass-access-token",
            refresh: "dev-bypass-refresh-token"
        )
        status = .signedIn
    }

    func signInWithBaseIdP() async {
        status = .signingIn
        errorMessage = nil
        do {
            let verifier = Self.randomURLSafeString(length: 48)
            let challenge = Self.s256Challenge(verifier)
            let stateParam = UUID().uuidString

            var components = URLComponents(
                url: APIConfig.idpIssuer.appendingPathComponent("oauth2/authorize"),
                resolvingAgainstBaseURL: false
            )!
            components.queryItems = [
                URLQueryItem(name: "response_type", value: "code"),
                URLQueryItem(name: "client_id", value: APIConfig.idpClientID),
                URLQueryItem(name: "redirect_uri", value: APIConfig.mobileRedirectURI),
                URLQueryItem(name: "scope", value: "openid profile"),
                URLQueryItem(name: "state", value: stateParam),
                URLQueryItem(name: "code_challenge", value: challenge),
                URLQueryItem(name: "code_challenge_method", value: "S256"),
            ]
            guard let authURL = components.url else {
                throw APIError.invalidURL
            }

            let callback = try await runWebAuth(url: authURL)
            guard let cb = URLComponents(url: callback, resolvingAgainstBaseURL: false),
                  let code = cb.queryItems?.first(where: { $0.name == "code" })?.value,
                  let returnedState = cb.queryItems?.first(where: { $0.name == "state" })?.value
            else { throw APIError.http(0, "Missing code in callback") }
            guard returnedState == stateParam else {
                throw APIError.http(0, "State mismatch")
            }

            struct ExchangeBody: Encodable {
                let code: String
                let code_verifier: String
                let redirect_uri: String
            }
            let resp: AuthResponse = try await APIClient.shared.post(
                "auth/idp-exchange",
                body: ExchangeBody(
                    code: code,
                    code_verifier: verifier,
                    redirect_uri: APIConfig.mobileRedirectURI
                )
            )
            storage.write(user: resp.user, access: resp.access_token, refresh: resp.refresh_token)
            status = .signedIn
        } catch {
            errorMessage = error.localizedDescription
            status = .error
        }
    }

    /// Dev fallback so the app is testable without registering this client in IdP.
    func signInWithEmail(email: String, password: String) async {
        status = .signingIn
        errorMessage = nil
        do {
            struct LoginBody: Encodable {
                let email: String
                let password: String
            }
            let resp: AuthResponse = try await APIClient.shared.post(
                "auth/login",
                body: LoginBody(email: email, password: password)
            )
            storage.write(user: resp.user, access: resp.access_token, refresh: resp.refresh_token)
            status = .signedIn
        } catch {
            errorMessage = error.localizedDescription
            status = .error
        }
    }

    func signOut() {
        storage.clear()
        status = .signedOut
    }

    // MARK: - PKCE helpers

    private static func randomURLSafeString(length: Int) -> String {
        let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_")
        var rng = SystemRandomNumberGenerator()
        return String((0..<length).map { _ in alphabet[Int(rng.next() % UInt64(alphabet.count))] })
    }

    private static func s256Challenge(_ verifier: String) -> String {
        let digest = SHA256.hash(data: Data(verifier.utf8))
        return Data(digest).base64URLEncoded()
    }

    // MARK: - ASWebAuthenticationSession bridge

    private func runWebAuth(url: URL) async throws -> URL {
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<URL, Error>) in
            let session = ASWebAuthenticationSession(
                url: url,
                callbackURLScheme: APIConfig.mobileRedirectScheme
            ) { callback, error in
                if let error = error {
                    cont.resume(throwing: error)
                } else if let callback = callback {
                    cont.resume(returning: callback)
                } else {
                    cont.resume(throwing: APIError.http(0, "No callback"))
                }
            }
            session.presentationContextProvider = self
            session.prefersEphemeralWebBrowserSession = false
            if !session.start() {
                cont.resume(throwing: APIError.http(0, "Failed to start auth session"))
            }
        }
    }
}

extension AuthController: ASWebAuthenticationPresentationContextProviding {
    nonisolated func presentationAnchor(for session: ASWebAuthenticationSession) -> ASPresentationAnchor {
        // Find the foreground active window scene.
        DispatchQueue.main.sync {
            UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .first(where: { $0.activationState == .foregroundActive })?
                .windows
                .first { $0.isKeyWindow } ?? ASPresentationAnchor()
        }
    }
}

private extension Data {
    func base64URLEncoded() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: client/lib/auth/auth_controller.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_web_auth_2/flutter_web_auth_2.dart';

import '../api/api_client.dart';
import '../api/api_config.dart';
import 'auth_storage.dart';

enum AuthStatus { unknown, signedOut, signingIn, signedIn, error }

class AuthState {
  const AuthState({
    required this.status,
    this.userName,
    this.userEmail,
    this.error,
  });

  final AuthStatus status;
  final String? userName;
  final String? userEmail;
  final String? error;

  AuthState copyWith({
    AuthStatus? status,
    String? userName,
    String? userEmail,
    String? error,
  }) =>
      AuthState(
        status: status ?? this.status,
        userName: userName ?? this.userName,
        userEmail: userEmail ?? this.userEmail,
        error: error,
      );

  static const initial = AuthState(status: AuthStatus.unknown);
}

class AuthController extends StateNotifier<AuthState> {
  AuthController(this._storage, this._api) : super(AuthState.initial) {
    _bootstrap();
  }

  final AuthStorage _storage;
  final ApiClient _api;

  Future<void> _bootstrap() async {
    final token = await _storage.readAccessToken();
    if (token == null || token.isEmpty) {
      state = state.copyWith(status: AuthStatus.signedOut);
      return;
    }
    final profile = await _storage.readProfile();
    state = state.copyWith(
      status: AuthStatus.signedIn,
      userName: profile['name'],
      userEmail: profile['email'],
    );
  }

  /// Dev-only: create a local-only "preview" session so the UI is browsable
  /// without hitting any auth server. Tokens are throwaway placeholders.
  Future<void> signInAsDev() async {
    state = state.copyWith(status: AuthStatus.signingIn, error: null);
    await _storage.writeSession(
      accessToken: 'dev-bypass-access-token',
      refreshToken: 'dev-bypass-refresh-token',
      userId: '00000000-0000-0000-0000-000000000000',
      email: 'preview@twende.local',
      fullName: 'Preview Traveler',
      role: 'traveler',
    );
    state = const AuthState(
      status: AuthStatus.signedIn,
      userName: 'Preview Traveler',
      userEmail: 'preview@twende.local',
    );
  }

  Future<void> signInWithBaseIdP() async {
    state = state.copyWith(status: AuthStatus.signingIn, error: null);
    try {
      final verifier = _generateVerifier();
      final challenge = _challenge(verifier);
      final stateParam = _randomString(24);

      final authorizeUrl = Uri.parse(
        '${ApiConfig.idpIssuer}/oauth2/authorize',
      ).replace(queryParameters: {
        'response_type': 'code',
        'client_id': ApiConfig.idpClientId,
        'redirect_uri': ApiConfig.mobileRedirectUri,
        'scope': 'openid profile',
        'state': stateParam,
        'code_challenge': challenge,
        'code_challenge_method': 'S256',
      }).toString();

      final result = await FlutterWebAuth2.authenticate(
        url: authorizeUrl,
        callbackUrlScheme: ApiConfig.mobileRedirectScheme,
      );

      final returned = Uri.parse(result);
      final code = returned.queryParameters['code'];
      final returnedState = returned.queryParameters['state'];

      if (returnedState != stateParam) {
        throw Exception('State mismatch — possible CSRF');
      }
      if (code == null || code.isEmpty) {
        throw Exception('Authorization code missing');
      }

      final resp = await _api.dio.post('/auth/idp-exchange', data: {
        'code': code,
        'code_verifier': verifier,
        'redirect_uri': ApiConfig.mobileRedirectUri,
      });

      final data = resp.data as Map<String, dynamic>;
      final user = data['user'] as Map<String, dynamic>;
      await _storage.writeSession(
        accessToken: data['access_token'] as String,
        refreshToken: data['refresh_token'] as String,
        userId: user['id'] as String,
        email: user['email'] as String,
        fullName: user['full_name'] as String? ?? '',
        role: user['role'] as String? ?? 'traveler',
      );
      state = AuthState(
        status: AuthStatus.signedIn,
        userName: user['full_name'] as String?,
        userEmail: user['email'] as String?,
      );
    } on DioException catch (e) {
      final msg = e.response?.data is Map
          ? (e.response!.data as Map)['error']?.toString() ?? e.message
          : e.message;
      state = state.copyWith(status: AuthStatus.error, error: msg);
    } catch (e) {
      state = state.copyWith(status: AuthStatus.error, error: e.toString());
    }
  }

  /// Dev-only email/password fallback so the app is testable without the IdP.
  Future<void> signInWithEmail(String email, String password) async {
    state = state.copyWith(status: AuthStatus.signingIn, error: null);
    try {
      final resp = await _api.dio.post('/auth/login', data: {
        'email': email,
        'password': password,
      });
      final data = resp.data as Map<String, dynamic>;
      final user = data['user'] as Map<String, dynamic>;
      await _storage.writeSession(
        accessToken: data['access_token'] as String,
        refreshToken: data['refresh_token'] as String,
        userId: user['id'] as String,
        email: user['email'] as String,
        fullName: user['full_name'] as String? ?? '',
        role: user['role'] as String? ?? 'traveler',
      );
      state = AuthState(
        status: AuthStatus.signedIn,
        userName: user['full_name'] as String?,
        userEmail: user['email'] as String?,
      );
    } on DioException catch (e) {
      final msg = e.response?.data is Map
          ? (e.response!.data as Map)['error']?.toString() ?? e.message
          : e.message;
      state = state.copyWith(status: AuthStatus.error, error: msg);
    } catch (e) {
      state = state.copyWith(status: AuthStatus.error, error: e.toString());
    }
  }

  Future<void> signOut() async {
    await _storage.clear();
    state = const AuthState(status: AuthStatus.signedOut);
  }

  String _generateVerifier() => _randomString(48);

  String _challenge(String verifier) =>
      base64Url.encode(sha256.convert(utf8.encode(verifier)).bytes).replaceAll('=', '');

  String _randomString(int len) {
    const alphabet =
        'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';
    final rng = Random.secure();
    return List.generate(len, (_) => alphabet[rng.nextInt(alphabet.length)]).join();
  }
}

final authControllerProvider =
    StateNotifierProvider<AuthController, AuthState>((ref) {
  final storage = ref.watch(authStorageProvider);
  final api = ref.watch(apiClientProvider);
  return AuthController(storage, api);
});
